src/coreclr/jit/helperexpansion.cpp

```diff
@@ -1553,3 +1602,317 @@ bool Compiler::fgVNBasedIntrinsicExpansionForCall_ReadUtf8(BasicBlock** pBlock,
+    {
+        // Possible scenario: someone changed Foo to be an interface,
+        // but static profile data still report it as a normal likely class.
+        JITDUMP("Likely class is abstract/interface - bail out (stale PGO data?).\n");
+        return false;
+    }
+
+    DebugInfo debugInfo = stmt->GetDebugInfo();
+
+    BasicBlock*    firstBb;
+    BasicBlock*    lastBb;
+    const unsigned tmpNum = SplitAtTreeAndReplaceItWithLocal(this, block, stmt, call, &firstBb, &lastBb);
+    lvaSetClass(tmpNum, expectedCls);
+
+    // Reload the arguments after the split
+    clsArg          = call->gtArgs.GetUserArgByIndex(0)->GetNode();
+    GenTree* objArg = call->gtArgs.GetUserArgByIndex(1)->GetNode();
+    *pBlock         = lastBb;
+
+    // We're going to expand this "isinst" like this:
+    //
+    // prevBb:                                      [weight: 1.0]
+    //     ...
+    //
+    // nullcheckBb (BBJ_COND):                      [weight: 1.0]
+    //     tmp = obj;
+    //     if (tmp == null)
+    //         goto lastBlock;
+    //
+    // typeCheckBb (BBJ_COND):                      [weight: 0.5]
+    //     if (tmp->pMT == likelyClass)
+    //         goto typeCheckSucceedBb;
+    //
+    // fallbackBb (BBJ_ALWAYS):                     [weight: <profile>]
+    //     tmp = helper_call(cls, tmp);
+    //     goto lastBlock;
+    //
+    // typeCheckSucceedBb (BBJ_ALWAYS):             [weight: <profile>]
+    //     nop (or tmp = null)
+    //
+    // lastBlock (BBJ_any):                         [weight: 1.0]
+    //     use(tmp);
+    //
+
+    // Block 1: nullcheckBb
+    // TODO-InlineCast: assertionprop should leave us a mark that objArg is never null, so we can omit this check
+    // it's too late to rely on upstream phases to do this for us (unless we do optRepeat).
+    GenTree* objTmp      = gtNewLclVarNode(tmpNum);
+    GenTree* nullcheckOp = gtNewOperNode(GT_EQ, TYP_INT, objTmp, gtNewNull());
+    nullcheckOp->gtFlags |= GTF_RELOP_JMP_USED;
+    BasicBlock* nullcheckBb = fgNewBBFromTreeAfter(BBJ_COND, firstBb, gtNewOperNode(GT_JTRUE, TYP_VOID, nullcheckOp),
+                                                   debugInfo, lastBb, true);
+
+    // Insert statements to spill call's arguments (preserving the evaluation order)
+    // TODO-InlineCast: don't spill cls if possible, we only need it after the nullcheck
+    const unsigned clsTmp   = lvaGrabTemp(true DEBUGARG("local for cls"));
+    lvaTable[clsTmp].lvType = clsArg->TypeGet();
+    Statement* storeObjStmt = fgNewStmtAtBeg(nullcheckBb, gtNewTempStore(tmpNum, objArg), debugInfo);
+    Statement* storeClsStmt = fgNewStmtAtBeg(nullcheckBb, gtNewTempStore(clsTmp, clsArg), debugInfo);
+    gtSetStmtInfo(storeObjStmt);
+    gtSetStmtInfo(storeClsStmt);
+    fgSetStmtSeq(storeObjStmt);
+    fgSetStmtSeq(storeClsStmt);
+    gtUpdateStmtSideEffects(storeObjStmt);
+    gtUpdateStmtSideEffects(storeClsStmt);
```

Comment:
This looks unnecessary. You've already split the statement at `call`, so its arguments are now going to be trivial/invariant nodes already.

Reply:
Removed completely as discussed 👍